engine/http.go

```diff
@@ -38,5 +40,20 @@ func (eng *Engine) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	// can only be sent before the body.
 	// (we could possibly use http footers via chunked encoding, but I couldn't find
 	// how to use them in net/http)
+
+	// Cancel the job if the HTTP connection is closed.
+	if closeNotifier, ok := w.(http.CloseNotifier); ok {
+		finished := make(chan struct{})
+		defer close(finished)
+		go func() {
+			select {
+			case <-closeNotifier.CloseNotify():
+				log.Infof("Client disconnected, cancelling job: %v", job)
```

Comment:
Good catch. This is actually a different interface which isn't usually used. I've forced to put it in the usual place too, which is here:

https://github.com/pwaller/docker/commit/d20a09e06ff1a933a49feab68f4150ac6fbec203#diff-44a527ea72ea9572e1cf71c7f1e2a0f0R1098